### ISSUE TITLE
Track the scraping status at the keyword level 

### ISSUE THREAD

**Issue by olivierobert** [liamstevens111/django-google-search-data-viewer#11]
## Issue

https://github.com/liamstevens111/google-search-data-viewer/blob/8ac9363bf1cffbfc063a5b9a05ebf1f8b6c9d04e/googlesearchdataviewer/apps/keywords/services.py#L18-L29

The scraping/processing status is managed at the upload level. Therefore, there is no handling for the failure of scraping at the keyword level i.e., some keywords might be correctly scraped while others might fail.

Note that there is also no background processing as you explained in your submission notes.

## Expected

While it is convenient for users to have a status to know that all keywords in an upload have been parsed, there should be a scraping status at the keyword level and the status at the upload should derive from the statuses of each keyword e.g., if all keywords in an uploaded succeeded then the status at the upload level should be "COMPLETED" but if the scraping failed for one more keywords, the status at the upload level should be "INCOMPLETE" or "FAILED". Then the user could see which ones failed and retry for those.

**Comment by liamstevens111** [comment]
Yes, I agree :100: It would make sense to have a status of "INCOMPLETE" for one failed keyword search and possibly even "FAILED" if they all did not complete.

Please let me know if you would like me to look into implementing this. I think I would need the following:

-  An Intermediary model/table named **Keyword** which would just hold the string value with no data. The original **KeywordResult** would then reference this as a Foreign Key and provide the additional data from the scrape if it completes.
- Remove **profile** from the **KeywordUploadProfile** model/table and rename it **KeywordUpload** and add an additional **status** field. The keyword would now reference the **Keyword** table holding the string and not the original **KeywordResult** as this would not have yet finished.

I could then allow the user to expand each of their uploads on the home page with a drop-down and show each keyword that was included in that upload and the individual status on the keyword level, even if it hasn't been searched/completed yet.

When the scrape finishes for all keywords in that upload I could then set the Upload status accordingly and create/update the final KeywordResult.

I look forward to hearing your thoughts. I could also provide a diagram of the design as it might help you understand my proposed design better than the wall of text above :man_facepalming:

**Comment by olivierobert** [comment]
Your rationale will suffice 😀  

Upon reviewing the codebase, the `profile` part was not very clear to me in all honesty 😅  So the renaming is welcome. As for separating the keyword from the keyword result, I would only recommend this implementation if there was a need to provide updated results for the same keyword (a keyword would `have_many` results). For simplicity, however, storing the keyword and result together would be more performant than having a relationship defined as a keyword `has_one` result. So it depends on how complex one wants to be.

**Comment by liamstevens111** [comment]
I think the Profile name appended to the end was due to the fact I included a Foreign Key to the Profile of the User who performed the upload so I could easily fetch the keywords that the user has searched without an extra query through the KeywordUpload table to find out the Profile. ie **KeywordUploadProfile.profile** instead of **KeywordUploadProfile.upload.uploader**. Not sure if that makes sense :sweat_smile:

Yes it would add complexity which would require joins etc, I think I thought about this idea to show to the user what keyword they searched and it's individual status so it could be stored and displayed independent of the result of scraping the keyword of Google; ie currently if the keyword scrape failed there would be **no** Keyword result stored in the database to show what was searched or to be able to show the user the current status.

**Comment by liamstevens111** [comment]
Had a look at the code again properly (I guess I forgot some reasoning and details since the weekend), I think the point of confusion is not the profile part of it but the upload field. 

Currently it is a ManyToMany / junction table holding the KeywordResult, Profile (user) and the Upload. The upload field is not being used at all but I initially planned to show the keywords searched under each Upload on the home page but as of right now it just shows the Upload and corresponding Status. I think I could delete this field and rename the model/table as KeywordSearchHistory, KeywordResultHistory or just KeywordProfile. Let me know what you think.
